### ISSUE TITLE
Typechecking doesn't support UDTs that reference other UDTs in the standard library

### ISSUE THREAD

**Issue by minestarks** [microsoft/qsharp#813]
This isn't an issue today *in practice*, because we don't have any UDTs in the stdlib that reference other UDTs.

But it is a bug in the compiler. It would be exposed if we started adding such types to the stdlib.

**Repro steps**

You need to update the stdlib to repro this bug. In `math.qs`, add the following type declaration:

```qsharp
newtype ComplexWrapper = (c: Complex);
```

Compile with `cargo build`.

Launch `target/debug/qsi` . Run the following Q#:

```qsharp
open Microsoft.Quantum.Math;
let c = Complex(1.0,1.0);
let x: Double = c::Real;
let cw = ComplexWrapper(c);
let cx: Complex = cw::c;
```

**Expected**

The last assignment should succeed.

**Actual**

```
error: Qsc.TypeCk.TyMismatch

  × type error
  ╰─▶ expected UDT<Item 210 (Package 1)>, found UDT<Item 210>
   ╭─[line_9:1:1]
 1 │ let cx: Complex = cw::c;
   ·                   ─────
   ╰────
```

Even though both sides of the assignment are type `Complex` , one of them is missing its `PackageId`, making the comparison fail.

Exists as of a153981617a6c8a2bb745a8d15176755d399af44

**Comment by minestarks** [comment]
FYI @swernli

**Comment by swernli** [comment]
Hmm... this seems like it could have a similar cause to the issue fixed in #766. I might be able to fix this using the same infra that introduced.

**Comment by minestarks** [comment]
@swernli I haven't confirmed this, but intuitively I also feel like this is related to the comments I left in my PR here:

https://github.com/microsoft/qsharp/pull/814/files#diff-2459af92d4e6b65ea27714c62d8ca0520a29f9e55c7148aef932b8eb8f2b7368R213

`typeck::Table::terms` doesn't keep track of the PackageId which makes it impossible to resolve UDTs through that type in the future. This is definitely causing bugs in the language service but I'm not sure if they're the same bug as this. For now, I'll assume the same, but we can revisit if they turn out to be separate bugs
